streams/src/main/java/org/apache/kafka/streams/state/internals/CachingSessionStore.java

```diff
@@ -170,7 +171,7 @@ private void putAndMaybeForward(final ThreadCache.DirtyEntry entry, final Intern
             final Bytes rawKey = Bytes.wrap(serdes.rawKey(key.key()));
             if (flushListener != null) {
                 final AGG newValue = serdes.valueFrom(entry.newValue());
-                final AGG oldValue = fetchPrevious(rawKey, key.window());
+                final AGG oldValue = sendOldValues ? fetchPrevious(rawKey, key.window()) : null;
```

Comment:
is there a reason why fetchPrevious has to rely on iterating over the store instead of just calling get() for the right key?

Reply:
There is no `get()` for a `SessionStore` The key in the session store is a combination of the record key, start and end time. We only know the start time for the previous key so we need to find the previous session with the correct start time.

Reply:
what about for window stores, couldn't we use get() there?

Reply:
No, it's a similar situation, there isn't a `get()` for a `WindowStore`.  The key is a combination of the record key and a timestamp when the record is placed in the store and records with the same key could be stored across multiple segments.  @dguy correct me if I'm mistaken here.

Reply:
I thought the timestamp would uniquely define the segment in which that key is stored.

Reply:
@xrvl there is no `get` on `WindowStore`. We could add one and it would work in scenarios where we don't have duplicates, i.e., the key for a WindowStore is (recordkey, timestamp, sequenceNumber) - if the store doesn't have duplicates the sequence number is always 0. If the store does have duplicates then we don't know what the sequence number is.
Without a KIP to add a `get()` to `WindowStore`, the only thing we could do is add a bit of a hack to see if the inner most store is a `RocksDBSegmentedBytesStore` and then we could call `get(..)` on that. If it isn't, then we'd still need to call `fetch`. 
For the DSL this would work as the only time we have duplicates in the `WindowStore` is for joins and we disable caching for those so it skips this code path. However, for the PAPI, we would need to always disable caching if duplicates are set. Which we probably should do anyway as it won't work as is.

Reply:
Thanks for the explanation @dguy, very helpful to understand where caching and sequence numbers come into play. It might be worthwhile to put this in a JIRA somewhere. I do think it would be a useful optimization to have eventually, as fetches have some setup / teardown overhead.